flink/src/main/java/org/apache/iceberg/flink/source/DataIterator.java

```diff
@@ -50,23 +54,36 @@
 abstract class DataIterator<T> implements CloseableIterator<T> {
 
   private Iterator<FileScanTask> tasks;
-  private final FileIO io;
-  private final EncryptionManager encryption;
+  private final Map<String, InputFile> inputFiles;
 
   private CloseableIterator<T> currentIterator;
 
   DataIterator(CombinedScanTask task, FileIO io, EncryptionManager encryption) {
     this.tasks = task.files().iterator();
-    this.io = io;
-    this.encryption = encryption;
+
+    Stream<EncryptedInputFile> encrypted = task.files().stream()
+        .flatMap(fileScanTask -> Stream.concat(Stream.of(fileScanTask.file()), fileScanTask.deletes().stream()))
+        .distinct()
```

Comment:
`distinct` compares files using `equals`, which is not overridden for data or delete files. This should instead use the approach that Spark uses:

```java
    Map<String, ByteBuffer> keyMetadata = Maps.newHashMap();
    task.files().stream()
        .flatMap(fileScanTask -> Stream.concat(Stream.of(fileScanTask.file()), fileScanTask.deletes().stream()))
        .forEach(file -> keyMetadata.put(file.path().toString(), file.keyMetadata()));
    Stream<EncryptedInputFile> encrypted = keyMetadata.entrySet().stream()
        .map(entry -> EncryptedFiles.encryptedInput(io.newInputFile(entry.getKey()), entry.getValue()));

    // decrypt with the batch call to avoid multiple RPCs to a key server, if possible
    Iterable<InputFile> decryptedFiles = encryptionManager.decrypt(encrypted::iterator);
```

Reply:
I see, updated.